File: app/cluster_objects/deployments.py
import app_config


def create_deployment(namespace: str, deployment_type: str, replicas: int):
    try:
        apps_v1 = app_config.apps_v1

        container_configs = {
            "app": {
                "name": "nginx",
                "image": "nginx:latest",
                "port": 80
            },
            "redis": {
                "name": "redis",
                "image": "redis:latest",
                "port": 6379
            },
            "mysql": {
                "name": "mysql",
                "image": "mysql:latest",
                "port": 3306,
                "env": [
                    {
                        "name": "MYSQL_ROOT_PASSWORD",
                        "value": "password123"
                    }
                ]
            }
        }

        config = container_configs[deployment_type]

        container = app_config.kubernetes_client.V1Container(
            name=config["name"],
            image=config["image"],
            ports=[app_config.kubernetes_client.V1ContainerPort(
                container_port=config["port"])],
            resources=app_config.kubernetes_client.V1ResourceRequirements(
                requests={
                    "cpu": "250m",
                    "memory": "128Mi"
                },
                limits={
                    "cpu": "500m",
                    "memory": "768Mi"
                }
            )
        )

        if deployment_type == "mysql" and "env" in config:
            container.env = [
                app_config.kubernetes_client.V1EnvVar(
                    name=env["name"], value=env["value"])
                for env in config["env"]
            ]

        deployment = app_config.kubernetes_client.V1Deployment(
            metadata=app_config.kubernetes_client.V1ObjectMeta(
                name=f"{deployment_type}-deployment",
                namespace=namespace,
                labels={"managed-by": "k8s-idp"}
            ),
            spec=app_config.kubernetes_client.V1DeploymentSpec(
                replicas=replicas,
                selector=app_config.kubernetes_client.V1LabelSelector(
                    match_labels={
                        "app": deployment_type,
                        "managed-by": "k8s-idp"
                    }
                ),
                template=app_config.kubernetes_client.V1PodTemplateSpec(
                    metadata=app_config.kubernetes_client.V1ObjectMeta(
                        labels={
                            "app": deployment_type,
                            "managed-by": "k8s-idp"
                        }
                    ),
                    spec=app_config.kubernetes_client.V1PodSpec(
                        containers=[container]
                    )
                )
            )
        )

        apps_v1.create_namespaced_deployment(
            namespace=namespace,
            body=deployment
        )

        return {
            "status": "success",
            "message": f"{deployment_type} deployment created successfully"
        }

    except Exception as e:
        return {
            "status": "error",
            "message": str(e)
        }


def get_existing_deployments(namespace: str):
    try:
        apps_v1 = app_config.apps_v1
        deployments = apps_v1.list_namespaced_deployment(
            namespace=namespace,
            label_selector="managed-by=k8s-idp"
        )

        existing = {}
        for deployment in deployments.items:
            deployment_type = deployment.metadata.name.replace(
                "-deployment", "")
            existing[deployment_type] = deployment.spec.replicas
        return existing

    except Exception as e:
        print(f"Error getting deployments: {e}")
        return {}


def scale_deployment(namespace: str, deployment_type: str, replicas: int):
    try:
        apps_v1 = app_config.apps_v1
        deployment_name = f"{deployment_type}-deployment"

        deployment = apps_v1.read_namespaced_deployment(
            name=deployment_name,
            namespace=namespace
        )

        deployment.spec.replicas = replicas

        apps_v1.patch_namespaced_deployment(
            name=deployment_name,
            namespace=namespace,
            body=deployment
        )

        return {
            "status": "success",
            "message": f"{deployment_type} scaled to {replicas} replicas"
        }

    except Exception as e:
        return {
            "status": "error",
            "message": str(e)
        }


def get_deployment_replicas(namespace: str, deployment_type: str):
    try:
        apps_v1 = app_config.apps_v1
        deployment_name = f"{deployment_type}-deployment"

        deployment = apps_v1.read_namespaced_deployment(
            name=deployment_name,
            namespace=namespace
        )

        return deployment.spec.replicas
    except Exception as e:
        print(f"Error getting replicas: {e}")
        return 1

File: internal/k8s/deployments.go
package k8s

import (
	"context"
	"fmt"
	"os"

	appsv1 "k8s.io/api/apps/v1"
	corev1 "k8s.io/api/core/v1"
	k8serrors "k8s.io/apimachinery/pkg/api/errors"
	"k8s.io/apimachinery/pkg/api/resource"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/utils/ptr"
)

// CreateDeployment creates a Kubernetes Deployment. If it exists, updates the replica count.
// TODO: Add service after first Deployment
func CreateDeployment(namespace, depType string, replicas int32) error {
	var containerName, image string
	var port int32
	var envs []corev1.EnvVar

	switch depType {
	case "app":
		containerName = "nginx"
		image = "nginx:latest"
		port = 80
	case "redis":
		containerName = "redis"
		image = "redis:latest"
		port = 6379
	case "mysql":
		containerName = "mysql"
		image = "mysql:latest"
		port = 3306
		// Read password from env
		pwd := os.Getenv("MYSQL_ROOT_PASSWORD")
		if pwd == "" {
			pwd = "password123" // Testing
		}
		envs = []corev1.EnvVar{
			{Name: "MYSQL_ROOT_PASSWORD", Value: pwd},
		}
	default:
		return fmt.Errorf("unsupported deployment type: %s", depType)
	}

	dep := &appsv1.Deployment{
		ObjectMeta: metav1.ObjectMeta{
			Name:      fmt.Sprintf("%s-deployment", depType),
			Namespace: namespace,
			Labels: map[string]string{
				"app":        depType,
				"managed-by": "k8s-idp",
			},
		},
		Spec: appsv1.DeploymentSpec{
			Replicas: ptr.To(replicas),
			Selector: &metav1.LabelSelector{
				MatchLabels: map[string]string{"app": depType},
			},
			Template: corev1.PodTemplateSpec{
				ObjectMeta: metav1.ObjectMeta{
					Labels: map[string]string{
						"app":        depType,
						"managed-by": "k8s-idp",
					},
				},
				Spec: corev1.PodSpec{
					Containers: []corev1.Container{
						{
							Name:  containerName,
							Image: image,
							Ports: []corev1.ContainerPort{{ContainerPort: port}},
							Env:   envs,
							Resources: corev1.ResourceRequirements{
								Requests: corev1.ResourceList{
									corev1.ResourceCPU:    resource.MustParse("250m"),  // TODO: parameterize them
									corev1.ResourceMemory: resource.MustParse("128Mi"), // TODO: parameterize them
								},
								Limits: corev1.ResourceList{
									corev1.ResourceCPU:    resource.MustParse("500m"),  // TODO: parameterize them
									corev1.ResourceMemory: resource.MustParse("768Mi"), // TODO: parameterize them
								},
							},
						},
					},
				},
			},
		},
	}

	appsClient := Client.AppsV1()
	ctx := context.Background()

	_, err := appsClient.Deployments(namespace).Create(ctx, dep, metav1.CreateOptions{})
	if err != nil {
		if k8serrors.IsAlreadyExists(err) {
			// BUG FIX: Instead of throwing 409 Conflict, we patch/update the existing deployment's replicas
			existing, getErr := appsClient.Deployments(namespace).Get(ctx, dep.Name, metav1.GetOptions{})
			if getErr != nil {
				return fmt.Errorf("failed to get existing deployment: %v", getErr)
			}
			existing.Spec.Replicas = ptr.To(replicas)
			_, updateErr := appsClient.Deployments(namespace).Update(ctx, existing, metav1.UpdateOptions{})
			if updateErr != nil {
				return fmt.Errorf("failed to update deployment replicas: %v", updateErr)
			}
			return nil
		}
		return err
	}

	return nil
}
